### DISCUSSION TITLE
How to handle multiple domains per MPI process

### DISCUSSION THREAD

**Discussion by benjaminjeliot** [Alpine-DAV/ascent#1105] [Q&A] [answered]
Hi everyone,

I was hoping someone could help me to understand how to handle multiple domains per process in Ascent.

I am currently trying to use Ascent in an app (very similar to the CloverLeaf 3D proxy) with 3D rectilinear mesh, multiple MPI processes and multiple domains per MPI process. Everything works fine with one domain per MPI process but fails when I try to use multiple domains per process.

- What do I need to change to handle multiple domains?
- How should I define the domain ID?
- How do I set multiple mesh domain data from a single process?
- How do I publish/execute multiple domains?

Many thanks,
Jim

**Comment by cyrush** [comment, marked as answer]
Hi Jim,

There is still one publish per MPI task, each MPI Task presents a conduit tree with children that represent the domains. 

Here is one way to create such a tree:

```
Node n;
Node & domain_0 = n.append();
// setup domain_0
Node & domain_1 = n.append();
// setup domain_1
```

For each domain, you can provide the domain id under `state/domain_id`:

```
domain_0["state/domain_id"] = 0;
```

Here are some test snippets that demo MPI use:

https://github.com/Alpine-DAV/ascent/blob/60dcf7d0a8de4e08f38aec92e13932f8cf88f145/src/tests/ascent/t_ascent_mpi_render_3d.cpp#L62

https://github.com/Alpine-DAV/ascent/blob/60dcf7d0a8de4e08f38aec92e13932f8cf88f145/src/tests/t_utils.hpp#L336


Hope this helps, and I will also admit we need better docs and examples that show the MPI case. 

-Cyrus

**Reply by benjaminjeliot** [reply]
Hi Cyrus,

Thanks for your reply - it made sense when you explained it.

It only took a few changes and now my code is working great. :-)

Kind regards,
Jim

**Reply by cyrush** [reply]
Glad that helped!
